fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -2041,40 +2047,128 @@ ACTOR Future<Void> TagPartitionedLogSystem::getDurableVersionChanged(LogLockInfo
 	return Void();
 }
 
+void getTLogLocIds(std::vector<Reference<LogSet>>& tLogs,
+                   std::vector<std::tuple<int, std::vector<TLogLockResult>>>& logGroupResults,
+                   std::vector<std::vector<int>>& tLogLocIds,
+                   int& maxTLogLocId) {
+	// Initialization.
+	tLogLocIds.clear();
+	tLogLocIds.resize(logGroupResults.size());
+	maxTLogLocId = std::numeric_limits<int>::min();
+
+	// Map the interfaces of all (local) tLogs to their corresponding locations in LogSets.
+	std::map<UID, int> interfLocMap;
+	int location = 0;
+	for (auto& it : tLogs) {
+		if (!it->isLocal) {
+			continue;
+		}
+		for (int i = 0; i < it->logServers.size(); i++) {
+			interfLocMap[it->logServers[i]->get().interf().id()] = location++;
+		}
+	}
+
+	// Find the locations of tLogs in "logGroupResults".
+	int logGroupId = 0;
+	for (auto& logGroupResult : logGroupResults) {
+		for (auto& tLogResult : std::get<1>(logGroupResult)) {
+			ASSERT(interfLocMap.find(tLogResult.logId) != interfLocMap.end());
+			tLogLocIds[logGroupId].push_back(interfLocMap[tLogResult.logId]);
+			maxTLogLocId = std::max(maxTLogLocId, interfLocMap[tLogResult.logId]);
+		}
+		logGroupId++;
+	}
+}
+
+void populateBitset(boost::dynamic_bitset<>& bs, std::vector<int>& ids) {
+	for (auto& id : ids) {
+		bs.set(id);
+	}
+}
+
 // If VERSION_VECTOR_UNICAST is enabled, one tLog's DV may advance beyond the min(DV) over all tLogs.
 // This function finds the highest recoverable version for each tLog group over all log groups.
 // All prior versions to the chosen RV must also be recoverable.
 // TODO: unit tests to stress UNICAST
-Version getRecoverVersionUnicast(std::vector<std::tuple<int, std::vector<TLogLockResult>>>& logGroupResults,
+Version getRecoverVersionUnicast(std::vector<Reference<LogSet>>& logServers,
+                                 std::vector<std::tuple<int, std::vector<TLogLockResult>>>& logGroupResults,
                                  Version minEnd) {
+	std::vector<std::vector<int>> tLogLocIds;
+	int maxTLogLocId = std::numeric_limits<int>::min();
+	getTLogLocIds(logServers, logGroupResults, tLogLocIds, maxTLogLocId);
+	int bsSize = maxTLogLocId + 1; // bitset size, used below
+
+	std::vector<Version> RVs(maxTLogLocId + 1, minEnd); // recovery versions of various tLogs
+
+	int tLogGroupIdx = 0;
 	Version minLogGroup = std::numeric_limits<Version>::max();
 	for (auto& logGroupResult : logGroupResults) {
```

Comment:
A comment about the recovery vector computation (not directly related to the changes in this PR): why are we computing the recovery version for each logGroup (which includes the satellite too) here? @dlambrig maybe you can explain this to me later, thanks!